File: OneToolkit.UI.Xaml/StringLocalizationConverter.cpp
#include "pch.h"

using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::UI::Xaml::Data;
using namespace Windows::UI::Xaml::Interop;
using namespace Windows::ApplicationModel::Resources;

namespace OneToolkit::UI::Xaml::Converters
{
	[Windows::Foundation::Metadata::WebHostHidden]
	public ref class StringLocalizationConverter sealed : IValueConverter, INotifyPropertyChanged
	{
	public:
		property ResourceLoader^ Context
		{
			ResourceLoader^ get()
			{
				return m_Context;
			}

			void set(ResourceLoader^ value)
			{
				if (m_Context != value)
				{
					m_Context = value;
					PropertyChanged(this, ref new PropertyChangedEventArgs("Context"));
				}
			}
		}

		virtual event PropertyChangedEventHandler^ PropertyChanged;

		virtual Object^ Convert(Object^ value, TypeName targetType, Object^ parameter, String^ language)
		{
			if (auto uri = dynamic_cast<Uri^>(value)) return m_Context->GetStringForUri(uri);
			else return m_Context->GetString(value->ToString());
		}

		virtual Object^ ConvertBack(Object^ value, TypeName targetType, Object^ parameter, String^ language)
		{
			throw ref new NotImplementedException();
		}
	private:	
		ResourceLoader^ m_Context;
	};
}

File: OneToolkit.UI.Xaml.Old/XamlConverters.Uwp/StringLocalizationConverter.cs
using System;
using Windows.ApplicationModel.Resources;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Data;

namespace OneToolkit.UI.Xaml.Converters
{
	public sealed class StringLocalizationConverter : DependencyObject, IValueConverter
	{
		public ResourceLoader Context
		{
			get => GetValue(ContextProperty) as ResourceLoader;
			set => SetValue(ContextProperty, value);
		}

		public static DependencyProperty ContextProperty { get; } = DependencyProperty.Register(nameof(Context), typeof(ResourceLoader), typeof(StringLocalizationConverter), new(GetDefaultContext()));

		private static ResourceLoader GetDefaultContext()
		{
			try { return ResourceLoader.GetForViewIndependentUse(); }
			catch { return null; }
		}


		public object Convert(object value, Type targetType, object parameter, string language) => Context.GetString(value.ToString());

		public object ConvertBack(object value, Type targetType, object parameter, string language) => throw new NotImplementedException();
	}
}